feat(lsp): add more LSP defaults (#28500)
neovim/neovim: runtime/doc/lsp.txt

  keymap exists
- |gr| is mapped in Normal mode to |vim.lsp.buf.references()| if no existing
  keymap exists
- CTRL-S is mapped in Insert mode to |vim.lsp.buf.signature_help()| if no
  existing keymap exists *i_CTRL-S*

                                                        *lsp-defaults-disable*
To override the above defaults, set or unset the options on |LspAttach|: >lua

    vim.api.nvim_create_autocmd('LspAttach', {
      callback = function(ev)
        vim.bo[ev.buf].formatexpr = nil
        vim.bo[ev.buf].omnifunc = nil
        vim.keymap.del('n', 'K', { buffer = ev.buf })
        vim.keymap.del('n', 'crr', { buffer = ev.buf })
        vim.keymap.del('n', 'cra', { buffer = ev.buf })
        vim.keymap.del('n', 'gr', { buffer = ev.buf })
        vim.keymap.del('i', '<C-S>', { buffer = ev.buf })
      end,
    })

To use other LSP features, set keymaps on |LspAttach|. Not all language
servers provide the same capabilities. To ensure you only set keymaps if the
language server supports a feature, guard keymaps behind capability checks.
Example: >lua

    vim.api.nvim_create_autocmd('LspAttach', {
      callback = function(args)

>         if client.supports_method(ms.textDocument_implementation) then

Reviewer: I love the use of the `Methods` module here; however I think there's still some discussion going on (see https://github.com/neovim/neovim/pull/24977) about its usage in examples.

Reviewer 2: I think some of the arguments made in that PR don't apply to the example here, but it looks like the `ms` definition is missing and that would definitely trip up people.

Author: Updated in ac59371ef9aa2c979043bee574e576bdf62fb664

  - To opt out of this use |gw| instead of gq, or set 'formatexpr' on LspAttach.
- |K| is mapped to |vim.lsp.buf.hover()| unless |'keywordprg'| is customized or
  a custom keymap for `K` exists.

> - *crr* is mapped in Normal mode to |vim.lsp.buf.rename()| if no existing

Reviewer: I think we should also document on how to modify/remove these if desired.

Author: Already mentioned just below this, but I updated the example to include the new default mappings too.

>   keymap exists
> - *cra* is mapped in Normal mode to |vim.lsp.buf.code_action()| if no existing
>   keymap exists
> - |gr| is mapped in Normal mode to |vim.lsp.buf.references()| if no existing
>   keymap exists
> - CTRL-S is mapped in Insert mode to |vim.lsp.buf.signature_help()| if no
>   existing keymap exists *i_CTRL-S*

Reviewer 3: ```suggestion
  mapping exists
- *cra* is mapped in Normal mode to |vim.lsp.buf.code_action()| if no existing
  mapping exists
- |gr| is mapped in Normal mode to |vim.lsp.buf.references()| if no existing
  mapping exists
- CTRL-S is mapped in Insert mode to |vim.lsp.buf.signature_help()| if no
  existing mapping exists *i_CTRL-S*
```

- |K| is mapped to |vim.lsp.buf.hover()| unless |'keywordprg'| is customized or
  a custom keymap for `K` exists.

                                                *gr-default*
                                                *i_CTRL-S*
Some keymaps are created unconditionally when Nvim starts:
- *crn* is mapped in Normal mode to |vim.lsp.buf.rename()|

> - *cra* is mapped in Normal mode to |vim.lsp.buf.code_action()|

Reviewer 2: ```suggestion
- *crr* is mapped in Normal and Visual mode to |vim.lsp.buf.code_action()|
```

(ignore the `crr` parts if the conclusion was to stick with `cra` - I don't mind which)

Author: I think the conclusion is to leave `crr` open for now, and use `crn` for rename and `cra` for code actions. @justinmk?

Reviewer 4: I'm in favor of `crr` for this PR. But I'm not going to block on that.

Author: Ok misunderstanding on my part, updated.

- |gr| is mapped in Normal mode to |vim.lsp.buf.references()|
- CTRL-S is mapped in Insert mode to |vim.lsp.buf.signature_help()|

If not wanted, these keymaps can be removed at any time using

> |vim.keymap.del()|.

Reviewer 4: ```suggestion
|vim.keymap.del()| or |:unmap|.
```

                                                        *lsp-defaults-disable*
To override the above defaults, set or unset the options on |LspAttach|: >lua

    vim.api.nvim_create_autocmd('LspAttach', {
      callback = function(ev)
        vim.bo[ev.buf].formatexpr = nil
        vim.bo[ev.buf].omnifunc = nil
        vim.keymap.del('n', 'K', { buffer = ev.buf })
      end,
    })

To use other LSP features, set keymaps on |LspAttach|. Not all language
servers provide the same capabilities. To ensure you only set keymaps if the
language server supports a feature, guard keymaps behind capability checks.
Example: >lua

    vim.api.nvim_create_autocmd('LspAttach', {
      callback = function(args)
        if client.supports_method('textDocument/implementation') then
          vim.keymap.set('n', 'g<C-I>', vim.lsp.buf.implementation, { buffer = args.buf })
        end

>       end,

Reviewer 4: Nice cleanup! I think `local client = vim.lsp.get_client_by_id(args.data.client_id)` is still needed though?

Author: Thanks, removed that by accident.
